src/typechecker/FStar.TypeChecker.Tc.fst

```diff
@@ -1100,14 +1100,23 @@ let finish_partial_modul (loading_from_cache:bool) (iface_exists:bool) (en:env)
 
   m, env
 
+let deep_compress_modul (m:modul) : modul =
+  { m with declarations = List.map (Compress.deep_compress_se false) m.declarations }
+
 let tc_modul (env0:env) (m:modul) (iface_exists:bool) :(modul * env) =
   let msg = "Internals for " ^ string_of_lid m.name in
   //AR: push env, this will also push solver, and then finish_partial_modul will do the pop
   let env0 = push_context env0 msg in
   let modul, env = tc_partial_modul env0 m in
+  // Compress all sigelts so we write a good checked file, plus we make
+  // sure that we are not leaking uvars, names, etc.
+  let modul = deep_compress_modul modul in
```

Comment:
I wonder if this compression can happen after type checking each sigelt, so that even when we add them to the typechecking environment, we know they are sane.

Reply:
Good point, let me try that.